src/vunnel/providers/rhel/csaf_client.py

```diff
@@ -0,0 +1,167 @@
+            for row in reader:
+                # row is like "2021/cve-2021-47265.json","2024-11-08T18:28:22+00:00"
+                changed_file = row[0]
+                date_str = row[1]
+                change_date = datetime.fromisoformat(date_str)
+                if self.archive_date and change_date < self.archive_date:
+                    break
+                if changed_file in seen_files:
+                    continue
+                seen_files.add(changed_file)
+        with concurrent.futures.ThreadPoolExecutor(max_workers=50) as executor:
+            futures = {
+                executor.submit(
+                    self._download_stream,
+                    url=self.latest_url.replace("archive_latest.txt", changed_file),
+                    path=os.path.join(self.advisories_path, changed_file),
+                ): changed_file
+                for changed_file in seen_files
+            }
+            concurrent.futures.wait(futures.keys())
+            for future, changed_file in futures.items():
+                if future.exception() is not None:
+                    self.logger.warning(f"Failed to download {changed_file}: {future.exception()}")
+
+    def _download_and_update_archive(self) -> None:
+        if not self.latest_archive_url:
+            self.latest_archive_url = self._archive_url()
+        archive_path = self._local_archive_path()
+        print(f"archive_path: {archive_path}")
+        if not os.path.exists(self.advisories_path):
+            os.makedirs(self.advisories_path)
+        # if there's a new one, the paths won't match and we need to download it
+        if not os.path.exists(archive_path):
+            self._download_stream(self._archive_url(), archive_path)
+            extract(archive_path, self.advisories_path)
+        # always download and process changes and deletions
+        # because this is how the API tells us about updates between
+        # publicatins of the main archive.
+        self._download_stream(self._changes_url(), os.path.join(self.workspace.input_path, "changes.csv"))
+        self._download_stream(self._deletions_url(), os.path.join(self.workspace.input_path, "deletions.csv"))
+        self.process_changes_and_deletions()
+
+    def path_from_rhsa_id(self, rhsa_id: RedHatAdvisoryID) -> str:
+        # https://security.access.redhat.com/data/csaf/v2/advisories/2024/rhba-2024_0599.json
+        return os.path.join(self.advisories_path, rhsa_id.advisory_year(), rhsa_id.advisory_id().lower().replace(":", "_") + ".json")
+
+    def csaf_doc_for_rhsa(self, rhsa: str) -> CSAFDoc | None:
+        """Get the CSAF document for a given RHSA ID"""
+        doc_path = self.path_from_rhsa_id(RedHatAdvisoryID(rhsa))
+        if os.path.exists(doc_path):
+            return _csaf_doc_from_path(doc_path)
+        self.logger.info(f"CSAF document not found for {rhsa}")
+        return None
+
+
+@functools.lru_cache(maxsize=1024)
+def _csaf_doc_from_path(doc_path: str) -> CSAFDoc:
```

Comment:
worth a comment on why this is useful performance-wise (since we're accessing the same doc multiple times)